doc/rst/language/spec/conversions.rst

```diff
@@ -12,79 +12,188 @@ expression can be a type expression. We refer to these two types the
+between two types, the compiler will check that a corresponding ``init=``
+also exists and emit an error if not.  Additionally, if ``init=`` is
+provided to initialize one type from another, the compiler will check
+that a corresponding ``:`` overload exists and will emit an error if not.
+See also :ref:`Explicit_Conversions` for more information on the ``:``
+operator. It is possible to provide ``:`` without ``init=`` or to provide
+``init=`` without ``=``.
+
+   *Example (implementing-assignment.chpl)*
+
+   Suppose that we have defined a record type to wrap an integer:
+
+   .. code-block:: chapel
+
+      record myInteger {
+        var intValue: int;
+      }
+
+   We might wish to support assignments setting a ``myInteger`` from
+   ``int``. In that event, we can provide the following functions:
+
+   .. code-block:: chapel
+
+      operator =(ref lhs: myInteger, rhs: int) {
+        lhs.intValue = rhs;
+      }
+      proc myInteger.init=(rhs: int) {
+        this.intValue = rhs;
+      }
+      operator :(from: int, type toType: myInteger) {
+        var tmp: myInteger = from; // invoke the init= above
+        return tmp;
+      }
+
+   All three of these functions are required if we wish to support
+   assignment. We can invoke these functions like this:
+
+   .. code-block:: chapel
+
+      var a = 1:myInteger;  // cast -- invokes operator :
+
+      var b: myInteger = 2; // initialization -- invokes init=
+
+      var c: myInteger;
+      c = 3;                // split-initialization -- invokes init=
+
+      var d = new myInteger();
+      d = 4;                // assignment -- invokes operator =
+
+   .. BLOCK-test-chapelnoprint
+
+      writeln("a is ", a, " : ", a.type:string);
+      writeln("b is ", b, " : ", b.type:string);
+      writeln("c is ", c, " : ", c.type:string);
+      writeln("d is ", d, " : ", d.type:string);
+
+   .. BLOCK-test-chapeloutput
+
+      a is (intValue = 1) : myInteger
+      b is (intValue = 2) : myInteger
+      c is (intValue = 3) : myInteger
+      d is (intValue = 4) : myInteger
+
+.. _Implicit_Conversion_Call:
+
+Implicit Conversions for Function Calls
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An implicit conversion for a function call - also called a *coercion* -
+occurs when the actual argument of a function call is converted to the
+type of the corresponding formal argument, if the formal’s intent is
+``param``, ``in``, ``const in``, or an abstract intent
+(:ref:`Abstract_Intents`) with the semantics of ``in`` or ``const in``.
+
+Additionally, the formal argument is considered to apply - without making
+a copy of the actual argument - when the actual type is a subtype of the
+formal type. This rule applies to ``in``, ``const in``, ``const ref``,
+and ``type`` intent formals and includes generic formal types.
+See:ref:`Subtype_Arg_Conversions`.
+
+Implicit conversions are not applied for actual arguments passed to
+``ref`` formal arguments.
```

Comment:
The above subsection needs to specify what conversions are allowed in this case, i.e. from what type to what type.

Subtyping is one of these allowed cases.

Note that when passing a subtype by `const ref`, the actual argument DOES need to be copied. So, I do not see the need to specify that the actual argument is not copied. Because in non-const-ref cases there is no observable difference.

Reply:
Right now we emit a compilation error when passing a subtype by `const ref` (that is not the formal type). It would not be reasonable to copy the actual for a `const ref` IMO because the identity would change but the `ref` part of the intent says it should not.

This issue issue #8489. However I think we are planning to support this eventually so I have been trying to describe the spec accordingly.

Reply:
But I see that the business about "without making a copy" seems to apply to the "in" intent in the way I had it, so I'll reword.